File: graphs & tries/2492.py
# GOOGLE
# Minimum score(distance) of path between two cities
# Undirected Graph
# 1 to n (1->4)
# DFS Traversal
# Tc = O(V + E)
# Sc = O(V + E)

from collections import defaultdict
import sys

class Solution:
    def dfs(self, u, adj, visited, result):
        visited.add(u)  # set to track visited nodes
        for v, d in adj[u]:
            result[0] = min(result[0], d)
            if v not in visited:
                self.dfs(v, adj, visited, result)

    def minScore(self, n, roads):
        adj = defaultdict(list)
        for u, v, d in roads:
            adj[u].append((v, d))
            adj[v].append((u, d))
        
        visited = set()  # set for visited nodes
        result = [sys.maxsize]  # list to allow modification in dfs
        self.dfs(1, adj, visited, result)
        return result[0]

File: graphs & tries/2492.js
class Solution {
    dfs(u, adj, visited, result) {
        visited.add(u);
        for (const [v, d] of adj.get(u)) {
            // Update the minimum score found so far
            result[0] = Math.min(result[0], d);
            if (!visited.has(v)) {
                this.dfs(v, adj, visited, result);
            }
        }
    }

    minScore(n, roads) {
        const adj = new Map();
        
        // Build the adjacency list
        for (const [u, v, d] of roads) {
            if (!adj.has(u)) adj.set(u, []);
            if (!adj.has(v)) adj.set(v, []);
            adj.get(u).push([v, d]);
            adj.get(v).push([u, d]);
        }
        
        const visited = new Set();
        const result = [Infinity]; // Initialize result with Infinity
        
        // Start DFS from node 1 (assuming nodes are 1-indexed)
        this.dfs(1, adj, visited, result);
        
        return result[0]; // Return the minimum score found
    }
}
